### PR TITLE
ci/caching: Unify (AZP) bazel build caches

### PR DESCRIPTION
**Pull request by phlax**
Currently all of our bazel jobs have their own caches - despite what is being cached is mostly the same

When jobs run without caches it lead to jobs failing throughout the CI run

This PR uses `bazel fetch` to create a bazel cache that all of the jobs can then use - and retries for flakey transient failures

Essentially the jobs are given 3 caches

- build image
- bazelisk - ie bazel itself
- bazel caches

on the basis that thse can expire independently

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by phlax** [comment]
i have been implementing something similar to the design here for AZP already with github workflows

i wasnt sure whether to apply the same here, but i think it will reduce a lot of CI flakes and make the bazel fetch part retriable so has some urgency

**Comment by phlax** [comment]
this also has quite a lot of additional benefits - for example we can make the scheduled run only run the cache warming env jobs - no need to run the full ci, and it will be _much_ quicker to warm caches and less likely to fail so should make that more reliable also

File: ci/format_pre.sh [hunk 1]
```diff
@@ -66,6 +66,10 @@ CURRENT=check_format
 echo "Running ${ENVOY_SRCDIR}/tools/code_format/check_format.py"
 time "${ENVOY_SRCDIR}/tools/code_format/check_format.py" fix --fail_on_diff
 
+ls -alh /build/.cache/bazelisk/downloads/bazelbuild
+# /bazelbuild/bazel-6.1.0-linux-arm64/bin
+ls -alh ~/.cache
```
Comment (phlax) [inline ci/format_pre.sh:71]: rm

File: ci/run_envoy_docker.sh [hunk 2]
```diff
@@ -65,7 +65,8 @@ fi
 
 if [[ -n "$ENVOY_DOCKER_PLATFORM" ]]; then
     echo "Setting Docker platform: ${ENVOY_DOCKER_PLATFORM}"
-    docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
+    PLATFORM_INSTALL="$(echo "${ENVOY_DOCKER_PLATFORM}" | cut -d/ -f2)"
```
Comment (phlax) [inline ci/run_envoy_docker.sh:68]: revert

File: .github/workflows/_env.yml [hunk 3]
```diff
@@ -152,6 +152,7 @@ jobs:
 
     - name: 'Print env'
       run: |
+        exit 1
```
Comment (phlax) [inline .github/workflows/_env.yml:155]: rm

File: .azure-pipelines/stage/publish.yml [hunk 4]
```diff
@@ -111,15 +111,16 @@ jobs:
       publishEnvoy: false
       publishTestResults: false
       # VERSION.txt is included to refresh Docker images for release
-      cacheKeyDocker: "ci/Dockerfile-envoy | VERSION.txt| $(cacheKeyBazelFiles)"
-      cacheKeyDockerName: publish_docker
+      cacheKeyDocker: "publish_docker.dev000 | ci/Dockerfile-envoy | VERSION.txt"
```
Comment (phlax) [inline .azure-pipelines/stage/publish.yml:114]:
```suggestion
      cacheKeyDocker: "publish_docker | ci/Dockerfile-envoy | VERSION.txt"
```

File: .azure-pipelines/stage/prechecks.yml [hunk 5]
```diff
@@ -59,6 +59,7 @@ jobs:
       publishEnvoy: false
       publishTestResults: false
       stepsPre:
+
```
Comment (phlax) [inline .azure-pipelines/stage/prechecks.yml:62]: revert

File: .azure-pipelines/docker/save_cache.sh [hunk 6]
```diff
@@ -28,6 +30,11 @@ if [[ -z "$NO_MOUNT_TMPFS" ]]; then
     mount -t tmpfs none "$DOCKER_CACHE_PATH"
 fi
 
+df -h
+
+du -sh /var/lib/docker
+du -sh "${DOCKER_CACHE_PATH}"
```
Comment (phlax) [inline .azure-pipelines/docker/save_cache.sh:36]: rm

File: .azure-pipelines/docker/prime_cache.sh [hunk 7]
```diff
@@ -14,27 +18,50 @@ else
...
 
-echo "Exporting /var/lib/docker -> ${DOCKER_CACHE_PATH}"
-mkdir -p "$DOCKER_CACHE_PATH"
-tar cf - -C /var/lib/docker . | zstd - -T0 -o "$DOCKER_CACHE_TARBALL"
+echo
+echo "================ Bazel fetch ======================"
+# Fetch bazel dependencies
+if [[ "$BAZEL_RESTORED" != "true" ]]; then
+    echo "Fetching Docker/bazel"
+    ./ci/run_envoy_docker.sh './ci/do_ci.sh fetch'
```
Comment (phlax) [inline .azure-pipelines/docker/prime_cache.sh:48]: else "not fetching ..."

**Comment by phlax** [comment]
this is another major change that fixes a few issues, and enables further related fixes/optimizations

there were some other changes i wanted to make immediately but held off as this is quite a big change already

**Comment by phlax** [comment]
testing this in other PRs and the retry works a treat

**Comment by phlax** [comment]
reWIPping to iron out some failures

**Review by yanavlasov — APPROVED** [review]

**Comment by phlax** [comment]
apologies for fp @yanavlasov i need to fix/test something and ended up doing a cleanup 

i will deWIP when i have tested further

**Comment by phlax** [comment]
@yanavlasov would you mind giving this another pass

File: .azure-pipelines/stages.yml [hunk 8]
```diff
@@ -57,6 +57,7 @@ stages:
   jobs:
   - template: env.yml
 
+
```
Comment (RyanTheOptimist) [inline .azure-pipelines/stages.yml:60]: Should this be removed?

**Review by RyanTheOptimist — APPROVED** [review]

File: .azure-pipelines/stages.yml (hunk 8 above)
Comment (phlax) [reply .azure-pipelines/stages.yml:60]: probably yes - i will do it in one of my follow up prs rather than running ci again for ws

**Pull request merged**